src/utils/command.rs

```diff
@@ -245,29 +271,82 @@ pub enum ParseError {
-            ParseError::UnknownCommand(e) => write!(f, "Unknown command: {}", e),
-            ParseError::WrongBotName(n) => write!(f, "Wrong bot name: {}", n),
-            ParseError::Custom(e) => write!(f, "{}", e),
+/// Command descriptions that can be shown to the user (e.g. as a part of
+/// `/help` message)
+///
+/// Most of the time you don't need to create this struct yourself as it's
+/// returned from [`BotCommands::descriptions`].
+#[derive(Debug, Clone)]
+pub struct CommandDescriptions<'a> {
+    global_description: Option<&'a str>,
+    descriptions: &'a [CommandDescription<'a>],
+    bot_username: Option<Cow<'a, str>>,
+}
+
+/// Description of a particular command, used in [`CommandDescriptions`].
+#[derive(Debug, Clone)]
+pub struct CommandDescription<'a> {
+    /// Prefix of the command, usually `/`.
+    pub prefix: &'a str,
+    /// The command itself, e.g. `start`.
+    pub command: &'a str,
+    /// Human-readable description of the command.
+    pub description: &'a str,
+}
+
+impl<'a> CommandDescriptions<'a> {
+    /// Creates new [`CommandDescriptions`] from a list of command descriptions.
+    pub fn new(descriptions: &'a [CommandDescription<'a>]) -> Self {
+        Self { global_description: None, descriptions, bot_username: None }
+    }
+
+    /// Sets the global description of these commands.
+    pub fn global_description(self, global_description: &'a str) -> Self {
+        Self { global_description: Some(global_description), ..self }
+    }
+
+    /// Sets the username of the bot.
+    ///
+    /// After this method is called, returned instance of
+    /// [`CommandDescriptions`] will append `@bot_username` to all commands.
+    /// This is useful in groups, to disambiguate commands for different bots.
+    ///
+    /// ## Examples
+    ///
+    /// ```
+    /// use teloxide::utils::command::{CommandDescription, CommandDescriptions};
+    ///
+    /// let descriptions = CommandDescriptions::new(&[
+    ///     CommandDescription { prefix: "/", command: "start", description: "start this bot" },
+    ///     CommandDescription { prefix: "/", command: "help", description: "show this message" },
+    /// ]);
+    ///
+    /// assert_eq!(descriptions.to_string(), "/start — start this bot\n/help — show this message");
+    /// assert_eq!(
+    ///     descriptions.username("username_of_the_bot").to_string(),
+    ///     "/start@username_of_the_bot — start this bot\n/help@username_of_the_bot — show this \
+    ///          message"
+    /// );
+    /// ```
+    pub fn username(self, bot_username: &'a str) -> Self {
+        Self { bot_username: Some(Cow::Borrowed(bot_username)), ..self }
+    }
+
+    /// Sets the username of the bot.
+    ///
+    /// This is the same as [`username`], but uses `get_me` method of the bot to
+    /// get the username.
+    ///
+    /// [`username`]: self::CommandDescriptions::username
+    pub async fn username_from_bot(self, bot: &impl Requester) -> CommandDescriptions<'a> {
```

Comment:
Is this function used somewhere?

Reply:
no

Reply:
What's the point then? If you already have a username, there is `username`:
https://github.com/teloxide/teloxide/blob/2c2be46561f9116491285033cf2c98bc5ef62f9a/src/utils/command.rs#L331

Reply:
You mean typing `.username(me.user.username.unwrap())`? Maybe accept `teloxide::types::Me` in `username` then and do the boilerplate inside the `username` function?

Reply:
Ah, `username` is optional there, uhh.

But what if you already just have the username? I think rather than changing `username` it would be best to add a different function for `Me` case. Like `me` or `username_from_me`.

Reply:
Okay, let's add `username_from_me` and invoke `username` from within. And regarding `username_from_bot`, I think unless this function is used somewhere, we'd better avoid it for now. We do already have a mechanism for getting `Me` here, and generally, I think it's better to just reuse it:

https://github.com/teloxide/teloxide/blob/dev/src/dispatching2/dispatcher.rs#L237

Reply:
Also, I suggest updating `examples/admin.rs` to use `username_from_me`. Rustdoc would inject this usage into the final docs of `username_from_me` and users will see how to use it in a real context.

Reply:
https://github.com/teloxide/teloxide/blob/f865f3a4d4c784ff7719259ce4783235308ffeb3/examples/dispatching2_features.rs#L138